File: src-lang/xt/db/node/proxy_base.clj
(ns xt.db.node.proxy-base
  (:require [hara.lang :as l]))

(l/script :xtalk
  {:require [[xt.lang.spec-base :as xt]
             [xt.lang.spec-promise :as promise]
             [xt.lang.common-data :as xtd]
             [xt.substrate :as substrate]
             [xt.substrate.page-proxy :as page-proxy]
             [xt.db.node.proxy-util :as proxy-util]]})

;;
;; Client-side proxy handlers for the server-side base adaptor handlers in
;; xt.db.node.kernel-base.
;;
;; These mirror the server substrate handler ids (e.g. @xt.db/init-base)
;; so that the same function ids can be invoked on a client node and forwarded
;; to the server over the configured transport.
;;
;; Substrate routing will always prefer an attached transport when no
;; transport_id is given, so proxy handlers must supply transport_id
;; explicitly from the request meta.
;;

(def.xt CALL_ACTIONS
  ["@xt.db/kernel-init"
   "@xt.db/kernel-setup"
   "@xt.db/kernel-teardown"
   "@xt.db/subscribe-db"
   "@xt.db/unsubscribe-db"
   "@xt.db/sync-cached"
   "@xt.db/rpc-call"
   "@xt.db/pull-call"
   "@xt.db/pull-cached"
   "@xt.db/dataview-call"
   "@xt.db/dataview-cached"])

(def.xt ATTACH_ACTIONS
  ["@xt.db/attach-model"
   "@xt.db/rpc-attach-model"
   "@xt.db/pull-attach-model"
   "@xt.db/dataview-attach-model"])

(def.xt DETACH_ACTIONS
  ["@xt.db/detach-model"])

(defn.xt request-proxy
  "generic client proxy for @xt.db/* actions"
  {:added "4.1"}
  [space args request node]
  (return (proxy-util/request-proxy space args request node)))

(defn.xt attach-forward-handler
  "forwards an attach action to the server and opens the proxy group"
  {:added "4.1"}
  [space args request node]
  (var page-args (xt/x:second args))
  (var space-id (xtd/get-in page-args ["space_id"]))
  (var group-id (xtd/get-in page-args ["group_id"]))
  (var transport-id (proxy-util/get-transport-id node (xtd/get-in request ["meta"])))
  (page-proxy/group-create-proxy node space-id group-id {} {"transport_id" transport-id})
  (return
   (promise/x:promise-then
    (substrate/request node
                       nil
                       (. request ["action"])
                       args
                       {"transport_id" transport-id})
    (fn [status]
      (return
       (promise/x:promise-then
        (page-proxy/group-open-proxy node space-id group-id {"transport_id" transport-id})
        (fn [_]
          (return status))))))))

(defn.xt detach-forward-handler
  "forwards a detach action to the server and closes the proxy group"
  {:added "4.1"}
  [space args request node]
  (var page-args (xt/x:second args))
  (var space-id (xtd/get-in page-args ["space_id"]))
  (var group-id (xtd/get-in page-args ["group_id"]))
  (var transport-id (proxy-util/get-transport-id node (xtd/get-in request ["meta"])))
  (return
   (-> (page-proxy/group-close-proxy node space-id group-id {"transport_id" transport-id})
       (promise/x:promise-then
        (fn [_]
          (return
           (substrate/request node
                              nil
                              (. request ["action"])
                              args
                              {"transport_id" transport-id})))))))

(defn.xt init-proxy-handlers
  "Registers client-side proxy handlers for xt.db.node.kernel-base methods so
   that the same substrate function ids can be invoked on a client node and
   forwarded to the server. Attach actions open a proxy group after the server
   call; detach actions close it."
  {:added "4.1"}
  [node]
  (xt/for:array [action -/CALL_ACTIONS]
    (substrate/register-handler node action -/request-proxy nil))
  (xt/for:array [action -/ATTACH_ACTIONS]
    (substrate/register-handler node action -/attach-forward-handler nil))
  (xt/for:array [action -/DETACH_ACTIONS]
    (substrate/register-handler node action -/detach-forward-handler nil))
  (return node))

File: packages-gen/dart/db/lib/node/proxy-base.dart
import 'package:xtalk_lang/common-data.dart' as xtd;

import 'package:xtalk_substrate/page-proxy.dart' as page_proxy;

import 'package:xtalk_db/node/proxy-util.dart' as proxy_util;

import 'package:xtalk_substrate/substrate.dart' as substrate;

var call_actions = <dynamic>[
  "@xt.db/kernel-init",
  "@xt.db/kernel-setup",
  "@xt.db/kernel-teardown",
  "@xt.db/subscribe-db",
  "@xt.db/unsubscribe-db",
  "@xt.db/sync-cached",
  "@xt.db/rpc-call",
  "@xt.db/pull-call",
  "@xt.db/pull-cached",
  "@xt.db/dataview-call",
  "@xt.db/dataview-cached"
];

var attach_actions = <dynamic>[
  "@xt.db/attach-model",
  "@xt.db/rpc-attach-model",
  "@xt.db/pull-attach-model",
  "@xt.db/dataview-attach-model"
];

var detach_actions = <dynamic>["@xt.db/detach-model"];

request_proxy(space, args, request, node) {
  return proxy_util.request_proxy(space,args,request,node);
}

attach_forward_handler(space, args, request, node) {
  var page_args = args[1];
  var space_id = xtd.get_in(page_args,<dynamic>["space_id"]);
  var group_id = xtd.get_in(page_args,<dynamic>["group_id"]);
  var transport_id = proxy_util.get_transport_id(node,xtd.get_in(request,<dynamic>["meta"]));
  page_proxy.group_create_proxy(
    node,
    space_id,
    group_id,
    <dynamic, dynamic>{},
    <dynamic, dynamic>{"transport_id":transport_id}
  );
  return ((Future.sync(() => substrate.request(
    node,
    null,
    request["action"],
    args,
    <dynamic, dynamic>{"transport_id":transport_id}
  ))) as Future<dynamic>).then((value) async { return await Function.apply((status) {
    return ((Future.sync(() => page_proxy.group_open_proxy(
      node,
      space_id,
      group_id,
      <dynamic, dynamic>{"transport_id":transport_id}
    ))) as Future<dynamic>).then((value) async { return await Function.apply((_) {
      return status;
    },<dynamic>[value]); });
  },<dynamic>[value]); });
}

detach_forward_handler(space, args, request, node) {
  var page_args = args[1];
  var space_id = xtd.get_in(page_args,<dynamic>["space_id"]);
  var group_id = xtd.get_in(page_args,<dynamic>["group_id"]);
  var transport_id = proxy_util.get_transport_id(node,xtd.get_in(request,<dynamic>["meta"]));
  return ((Future.sync(() => page_proxy.group_close_proxy(
    node,
    space_id,
    group_id,
    <dynamic, dynamic>{"transport_id":transport_id}
  ))) as Future<dynamic>).then((value) async { return await Function.apply((_) {
    return substrate.request(
      node,
      null,
      request["action"],
      args,
      <dynamic, dynamic>{"transport_id":transport_id}
    );
  },<dynamic>[value]); });
}

init_proxy_handlers(node) {
  var arr_43084 = call_actions;
  for(var i43085 = 0; i43085 < arr_43084.length; ++i43085){
    var action = arr_43084[i43085];
    substrate.register_handler(node,action,request_proxy,null);
  };
  var arr_43106 = attach_actions;
  for(var i43107 = 0; i43107 < arr_43106.length; ++i43107){
    var action = arr_43106[i43107];
    substrate.register_handler(node,action,attach_forward_handler,null);
  };
  var arr_43128 = detach_actions;
  for(var i43129 = 0; i43129 < arr_43128.length; ++i43129){
    var action = arr_43128[i43129];
    substrate.register_handler(node,action,detach_forward_handler,null);
  };
  return node;
}